### ISSUE TITLE
Add support for streams

### ISSUE THREAD

**Issue by tirkarthi** [mitsuhiko/redis-rs#162]
Redis 5.0 will ship with streams. It's in beta now and will be stable in few months. It might be a good addition to the library and I hope it will work well with async libraries in Rust. I couldn't see any related issues or PR. Feel free to close this if it's a duplicate.

Streams reference : https://redis.io/topics/streams-intro

Thanks much for the library!

**Comment by badboy** [comment]
I played around with prototyping a stream API when streams were developed initially (I never published or even finished that though).
I'm convinced we can provide a really good API to make this all work, however I personally don't have a use case for streams on Redis currently, so I won't spend time on developing such an API.
I'd be happy to guide external contributors in implementing that though.

**Comment by lambdasqd** [comment]
I would be interested in giving this a try later this week.

**Comment by grippy** [comment]
I've completed most of the commands for Redis Streams. I'm currently working my way through the mountain of tests for this change. And I'm saving the docs for last once everything checks out.

Since there are various commands, which all support a mix of fixed positional arguments + optional args, I've had to add a few builders for constructing the optional args. The current lib doesn't offer much in guidance on the preferred way to implement optional args. I tried to avoid them and use variations of the same redis cmd to denote the difference. This helps but I just couldn't tackle the `xread` and `xclaim` without the optionals. 

Similarly, there's at least three or four complex response types (think InfoDict but not clean as wrapping a HashMap). So, I've added a bunch of new code. This turned out much larger then what I had originally hoped for. However, adding a partial set of stream commands doesn't make this useful. 

I currently have everything structured as follows:

- All the new Stream* types are added to the end of `types.rs` 
- All the new commands are added alphabetically to `commands.rs`
- I created a new file called `tests/test_streams.rs` for test isolation.

Does this structure work? Or, is there some other way to structure this change that makes more sense (like, maybe move all the new types into their own module)? I don't have a preference. Right now, all the new types are prefixed with `Stream.` It feels like I could remove this if they end up in their own module.

Any guidance on the structure would be helpful.

**Comment by badboy** [comment]
How much does it really depend on being inside redis-rs?
I'm not very keen of pulling a bunch of code into this library and needing to maintain it (as I don't have any personal use for streams).

IMO, as mentioned before, I think it might be possible to build this as its own crate on top of redis-rs.
That would also directly solve the code organization problem by being its own namespace.

**Comment by grippy** [comment]
@badboy Currently, these commands are being added to the client inside the `implement_commands` macro. This seems logical (to me) since a client automatically picks up all the commands inside the Commands trait. Unless there was an existing plugin architecture for implementing commands (either standalone or grouped), I don't know how I could add new commands directly to the client connection in a separate crate.

You're correct, this could definitely be done in a sep crate. So could the Geo commands (and any other new commands added down the road). But now each of these new crates needs to implement connection handling (sync/async) (as well other boilerplate code.. I need to grasp what exactly). Now if you want to use Streams, Geo, and any of the currently supported functionality, you're looking at three different crates and three different clients (two wrappers + the redis-rs client). 

I completely understand not wanting to maintain something you don't currently have the need for. I myself don't currently have a use case for Streams. But that doesn't mean someone else doesn't. Anyway, I'm completely open to helping out with whatever future changes/bugs need to be added here. Personally, I would rather do that under a project that has a community & maintainers than off on my own without anyone to help/guide code reviews.

**Comment by badboy** [comment]
> Unless there was an existing plugin architecture for implementing commands (either standalone or grouped), I don't know how I could add new commands directly to the client connection in a separate crate.

So there's no direct requirement to be in redis-rs. There's no plugin mechanism, because all of this can already be done using traits, e.g. a `trait StreamCommands` in your own crate.

> Now if you want to use Streams, Geo, and any of the currently supported functionality, you're looking at three different crates and three different clients (two wrappers + the redis-rs client).

Then again in Rust crates are the way to go. And we're not anywhere close to having _too_ many crates for doing Redis.

> I myself don't currently have a use case for Streams. But that doesn't mean someone else doesn't.

I don't doubt that people might have a use for it, but I don't have the capacity to do the work for them now.

> Personally, I would rather do that under a project that has a community & maintainers than off on my own without anyone to help/guide code reviews.

But so far you don't have code reviewers here either. Again: I don't have the bandwidth to do it and currently am simply not interested in reviewing stream code.
IMO a separate crate to test it out & figuring out a good API is the way to go. That doesn't preclude that in the future redis-rs might merge the code.

**Comment by Marwes** [comment]
I think it is fair to do a PR with the code as is (integrated in this) repo. I'd be happy to look at it and it will be easier to point out where it can be separated to its own crate that way also.

**Comment by badboy** [comment]
Thanks, @Marwes!

**Comment by grippy** [comment]
@Marwes Thanks. I agree, this will a good starting point for some feedback. Then let's see where it goes from there. 

I just finished wrapping up the tests. Will do a pass at the docs next and then submit a PR.

**Comment by grippy** [comment]
@badboy I took your advice and moved my changes into a new crate `redis-streams-rs`. After experimenting with a PING command, I didn't realize how easy this would be. Well done.

@Marwes If the offer still stands, I would love some feedback. There's a lot of new code here:
https://github.com/grippy/redis-streams-rs/pull/1

Also, if anyone has time, my new crate has a dependency on this PR:
https://github.com/mitsuhiko/redis-rs/pull/203

**Comment by application-developer-DA** [comment]
@grippy is it planned to keep the implementation inside the `redis-stream-rs` crate or do you plan to make a PR eventually, so that the implementation of `redis-stream-rs` gets merged into `redis-rs`?

**Comment by grippy** [comment]
@application-developer-DA I'm open to folding it into this crate as a feature flag (if that makes sense). However, @badboy expressed the desire for this to live as a standalone crate.

**Comment by application-developer-DA** [comment]
I'm just thinking that some changes are better to be done at both places, like for instance the implementation of `Commands`, `StreamCommands` for non-blocking version of `Connection` (for tokio/futures usage). Also I've already encountered a problem that when I depend on both `redis-rs` and `redis-streams-rs`, then I can't use the `StreamCommands` trait anymore as it thinks that a `Connection` from `redis-rs` and from `redis-stream-rs` are 2 different types (I believe that's because `redis-stream-rs` uses a master version of `redis-rs`, whereas the version from `crates.io.` is a bit older), so normally it forces the user to check the dependencies and "resolve this conflict" manually by making sure that both rely from the same library.

One could have made the `redis` available from `redis-rs` as `pub use`, so that everyone who wants to do something with `redis-rs`, should use the `redis_streams::redis::...` syntax, but I'm not sure if this is the most optimal strategy, especially if `redis-rs` will be updated more often (in this case the one might want to use thee latest version of `redis-rs`).

**Comment by ilovelll** [comment]
Any update on this issue?

**Comment by Terkwood** [comment]
I find it strange to use two crates for this. Streams commands are a first class abstraction in Redis since 5.0 (not a an optional, module-based subsystem), so it seems odd to segment them off into a separate crate. The current low-level functionality offered by `redis-rs` is certainly enough to manage most (maybe all) use-cases, but it's verbose in comparison to the high-level interface provided for older commands like get/set/hgetall.

The contributed crate from from @grippy has a higher-level API which looks easy to use. It currently needs a [small update](https://github.com/grippy/redis-streams-rs/issues/12) to move its `redis` dependency forward.

**Comment by grippy** [comment]
@Terkwood Thanks for the PR. I agree with you and @application-developer-DA that it isn't optimal having this as a separate lib because of the issues mentioned above.

@badboy @Marwes What do you think about folding the streams commands into this project similar to how the `geo` commands were handled? I haven't seen many reported issues on the streams lib since it was released.

**Comment by badboy** [comment]
> @badboy @Marwes What do you think about folding the streams commands into this project similar to how the `geo` commands were handled? I haven't seen many reported issues on the streams lib since it was released.

I'm not super oposed to merging it. If you send a PR I can put it on my list to review, I can't guarantee to give it a look soon though

**Comment by Terkwood** [comment]
I'm still very slowly reviewing the new Redis Streams work (notably by making real use of it in a hobbyist software system) and hope to provide constructive feedback based on that experience.

After that - - if no one else beats me to it - - I'll be glad to start drafting the PR for the merge of the new Redis Streams lib into _this_ code base.

Thanks again for all the collaboration on this thread.

**Comment by Terkwood** [comment]
I've drafted some work at https://github.com/Terkwood/redis-rs/pull/1, if anyone in this thread wants an early preview. 

Once the draft is complete, I'll open a separate PR against this repository for final review, in order to make the best use of @badboy's time.  Cheers!

**Comment by Terkwood** [comment]
The bulk of the work for merging https://github.com/grippy/redis-streams-rs into this repo is complete in PR #319.

There are [a few outstanding lints](https://travis-ci.org/github/mitsuhiko/redis-rs/jobs/688571409) that I'll address presently.  Until that's done, I'll leave the PR marked as a draft.  (But it's more or less ready for review by whoever is interested.)

**Comment by Terkwood** [comment]
This is very nearly complete in #319.  There's an additional pull request related to the `FromRedisValue` conversion for tuples (#328) which needs to be merged before #319 can be merged, so that #319 avoids an extraneous round of copies.

**Comment by Terkwood** [comment]
Anyone who's interested:

Please vote for #319! 🗳 

It's ready for a merge! 😁😅😂

**Comment by Terkwood** [comment]
Thanks again to all the contributors in this effort! Streams support is now available in `master` branch. 🔥 🏞

**Comment by badboy** [comment]
Thanks all! (and sorry for not being able to spend time on it, I'm glad we have enough contributors now that can push this forward).
